### PR TITLE
Add support for <fullTestName>_SET_RUN_SERIAL=[ON|OFF] (trilinos/Trilinos#7112)

### PR DESCRIPTION
**Pull request by bartlettroscoe**
This PR implements support for the option `<fullTestName>_SET_RUN_SERIAL=[ON|OFF]` described in trilinos/Trilinos#7112.  This will be used on the Trilinos side immediately to set `RUN_SERIAL` on several Trilinos tests that are timing out in one of the builds on 'ride'.  (I tested and setting `RUN_SERIAL` eliminates these timeouts.  I will document that testing more completely in the matching Trilinos PR shortly.)

## How was this tested?

There are strong unit tests for this and, again, I validated this against Trilinos.  I used this TriBITS branch with a matching Trilinos branch to run the full build `Trilinos-atdm-white-ride-gnu-7.2.0-openmp-debug-exp` which posted to CDash:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&date=2020-09-18&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-white-ride-gnu-7.2.0-openmp-debug-exp&field2=buildstamp&compare2=61&value2=20200918-2241-Experimental

Looking at the [configure results](https://testing.sandia.gov/cdash/build/7904253/configure) showed:

```
-- Setting default Intrepid2_unit-test_Projection_OpenMP_Test_Convergence_HEX_MPI_1_SET_RUN_SERIAL=ON
....
-- Intrepid2_unit-test_Projection_Serial_Test_Convergence_HEX_MPI_1: Added test (BASIC, NUM_MPI_PROCS=1, PROCESSORS=1, RUN_SERIAL)!
```

and the [cache file](https://testing.sandia.gov/cdash/build/7904253/notes#note5) showed:

```
Intrepid2_unit-test_Projection_OpenMP_Test_Convergence_HEX_MPI_1_SET_RUN_SERIAL:BOOL=ON
```

Again, more validation evidence will be given in the matching Trilinos PR that will be created shortly.

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
@jmgate, any time to review this one?

File: test/core/TestingFunctionMacro_UnitTests.cmake [hunk 1]
```diff
@@ -978,6 +995,72 @@ FUNCTION(UNITTEST_TRIBITS_ADD_TEST_BASIC)
...
+  UNITTEST_COMPARE_CONST( ${EXEN}_TEST_NAMES
+    "PackageA_SomeExec" )
+  UNITTEST_COMPARE_CONST(
+    TRIBITS_ADD_TEST_ADD_TEST_INPUT
+    "${PACKEXEN};${PACKEXEN_PATH}"
+    )
+  UNITTEST_COMPARE_CONST( MESSAGE_WRAPPER_INPUT
+    "-- PackageA_SomeExec: Added test (BASIC, PROCESSORS=1)!" )
+  UNITTEST_NOT_HAS_SUBSTR_CONST(TRIBITS_SET_TEST_PROPERTIES_INPUT
+    "RUN_SERIAL")
```
Comment (e10harvey) [inline test/core/TestingFunctionMacro_UnitTests.cmake:1062]: `RUN_SERIAL;OFF`?

File: test/core/TestingFunctionMacro_UnitTests.cmake [hunk 2]
```diff
@@ -978,6 +995,72 @@ FUNCTION(UNITTEST_TRIBITS_ADD_TEST_BASIC)
...
+  UNITTEST_COMPARE_CONST( ${EXEN}_TEST_NAMES
+    "PackageA_SomeExec" )
+  UNITTEST_COMPARE_CONST(
+    TRIBITS_ADD_TEST_ADD_TEST_INPUT
+    "${PACKEXEN};${PACKEXEN_PATH}"
+    )
+  UNITTEST_COMPARE_CONST( MESSAGE_WRAPPER_INPUT
+    "-- PackageA_SomeExec: Added test (BASIC, PROCESSORS=1, RUN_SERIAL)!" )
+  UNITTEST_HAS_SUBSTR_CONST(TRIBITS_SET_TEST_PROPERTIES_INPUT
+    "RUN_SERIAL")
```
Comment (e10harvey) [inline test/core/TestingFunctionMacro_UnitTests.cmake:1045]: `RUN_SERIAL;OFF`?

File: test/core/TestingFunctionMacro_UnitTests.cmake [hunk 3]
```diff
@@ -978,6 +995,72 @@ FUNCTION(UNITTEST_TRIBITS_ADD_TEST_BASIC)
...
```
Comment (e10harvey) [inline test/core/TestingFunctionMacro_UnitTests.cmake:1012]: `RUN_SERIAL;ON`?

**Review by e10harvey — COMMENTED** [review]
In general, everything looks great. I do have some questions about the unit tests; I want to confirm that `-DTEST_NAME_SET_RUN_SERIAL=OFF` works.

File: test/core/TestingFunctionMacro_UnitTests.cmake (hunk 1 above)
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:1062]: If the `RUN_SERIAL` property is not set, it defaults to OFF.  IMHO, it is better to not set it as `RUN_SERIAL OFF` in this case.

File: test/core/TestingFunctionMacro_UnitTests.cmake (hunk 2 above)
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:1045]: No, this should look for `RUN_SERIAL;ON` in this case just to be sure.  I will create a new TriBITS PR that looks for `RUN_SERIAL;ON` in this check.

File: test/core/TestingFunctionMacro_UnitTests.cmake (hunk 3 above)
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:1012]: No, this should look for `RUN_SERIAL;ON` in this case just to be sure.  I will create a new TriBITS PR that looks for `RUN_SERIAL;ON` in this check.

File: test/core/TestingFunctionMacro_UnitTests.cmake [hunk 4]
```diff
@@ -3548,6 +3706,39 @@ FUNCTION(UNITTEST_TRIBITS_ADD_ADVANCED_TEST_EXCLUDES)
...
+  MESSAGE("***\n*** Set RUN_SERIAL argument\n***")
+  TRIBITS_ADD_ADVANCED_TEST_UNITTEST_RESET()
+  SET(${PACKAGE_NAME}_TAAT_mpi_cmnd_0_cmnd_1_SET_RUN_SERIAL "")
+  TRIBITS_ADD_ADVANCED_TEST( TAAT_mpi_cmnd_0_cmnd_1 RUN_SERIAL
+    TEST_0 CMND someCmnd TEST_1 CMND someOtherCmnd )
+  UNITTEST_COMPARE_CONST( MESSAGE_WRAPPER_INPUT
+    "-- PackageA_TAAT_mpi_cmnd_0_cmnd_1: Added test (BASIC, PROCESSORS=1, RUN_SERIAL)!" )
+  UNITTEST_COMPARE_CONST(TRIBITS_ADD_ADVANCED_TEST_NUM_CMNDS "2")
+  UNITTEST_HAS_SUBSTR_CONST(TRIBITS_SET_TEST_PROPERTIES_INPUT
+    "RUN_SERIAL;ON")
```
[the remark this answers is not in the document]
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:3729]: If the `RUN_SERIAL` property is not set, it defaults to OFF.  IMHO, it is better to not set it as `RUN_SERIAL OFF` in this case.

File: test/core/TestingFunctionMacro_UnitTests.cmake [hunk 5]
```diff
@@ -3548,6 +3706,39 @@ FUNCTION(UNITTEST_TRIBITS_ADD_ADVANCED_TEST_EXCLUDES)
...
+  MESSAGE("***\n*** Set RUN_SERIAL input argument but set <fullTestName>_SET_RUN_SERIAL=OFF\n***")
+  TRIBITS_ADD_ADVANCED_TEST_UNITTEST_RESET()
+  SET(${PACKAGE_NAME}_TAAT_mpi_cmnd_0_cmnd_1_SET_RUN_SERIAL OFF)
+  TRIBITS_ADD_ADVANCED_TEST( TAAT_mpi_cmnd_0_cmnd_1 RUN_SERIAL
+    TEST_0 CMND someCmnd TEST_1 CMND someOtherCmnd )
+  UNITTEST_COMPARE_CONST( MESSAGE_WRAPPER_INPUT
+    "-- PackageA_TAAT_mpi_cmnd_0_cmnd_1: Added test (BASIC, PROCESSORS=1)!" )
+  UNITTEST_COMPARE_CONST(TRIBITS_ADD_ADVANCED_TEST_NUM_CMNDS "2")
+  UNITTEST_NOT_HAS_SUBSTR_CONST(TRIBITS_SET_TEST_PROPERTIES_INPUT
+    "RUN_SERIAL;ON")
```
[the remark this answers is not in the document]
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:3740]: No, we want to make sure that `RUN_SERIAL` is not set to `ON` due to the desired behavior documented in:

* https://github.com/TriBITSPub/TriBITS/pull/328/files/32e3fbc0ead0903c7e3ab48a1c5819544490e964#diff-251eb5983dbc30ed89c9d0fa1b91bb5cR2061

I will change this to make sure that `RUN_SERIAL` is not set at all.

**Comment by bartlettroscoe** [comment]
@e10harvey, thanks for the helpful review.  I think the updated unit tests will be better given you feedback.  (I will post a follow-on PR for those fixes.)

**Comment by jmgate** [comment]
Sorry, @bartlettroscoe, I didn't get a notification again.  I've tweaked my notification settings one more time to see if the emails finally start coming in.

File: test/core/TestingFunctionMacro_UnitTests.cmake (hunk 5 above)
[the remark this answers is not in the document]
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:3740]: Addressed in TriBITSPub/TriBITS#330

File: test/core/TestingFunctionMacro_UnitTests.cmake (hunk 2 above)
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:1045]: Addressed in TriBITSPub/TriBITS#330

File: test/core/TestingFunctionMacro_UnitTests.cmake (hunk 3 above)
Comment (bartlettroscoe) [reply test/core/TestingFunctionMacro_UnitTests.cmake:1012]: Addressed in TriBITSPub/TriBITS#330

**Comment by jmgate** [comment]
@bartlettroscoe, I'm remembering from a long time ago that there are occasionally issues that come up with GitHub notification emails not getting through to Sandia email addresses.  Do you recall any details of that problem, or what the solution was?

**Pull request merged**